### PR TITLE
adding code to generate hexadecimal numbers in java tests

### PR DESCRIPTION
**Pull request by eigold**

### CODE REVIEW

File: src/goto-programs/interpreter.cpp [hunk 1]
```diff
@@ -960,7 +960,8 @@ bool interpretert::unbounded_size(const typet &type)
 size_t interpretert::get_size(const typet &type)
 {
   if(unbounded_size(type))
-    return 2ULL << 32ULL;
+    return 1ULL << 31ULL;
+    //return 2ULL << 32ULL;
```
Comment (martin-cs) [inline src/goto-programs/interpreter.cpp]: This seems unrelated to the actual change.  It's also not clear why it's needed.

File: src/java_bytecode/expr2java.cpp [hunk 2]
```diff
@@ -235,6 +237,23 @@ std::string expr2javat::convert_constant(
...
+  }
+   else if (src.type() == java_double_type()) {
+    ieee_floatt Fpn(to_constant_expr(src));
+    double D = Fpn.to_double();
+    std::stringstream buffer;
+    buffer << std::hexfloat;
+    buffer << D;
+    return(buffer.str());
+  }
+
```
Comment (martin-cs) [inline src/java_bytecode/expr2java.cpp]: Seems a bit clunky but I think that's just what C++ forces you to do.  Other than that; looks good.

File: src/goto-programs/interpreter.cpp (hunk 1 above)
Comment (mgudemann) [inline src/goto-programs/interpreter.cpp]: this is required for 32 bit systems, as the original code is UB for those, it's not yet obvious how to fix this in general for 32 bit
Still, it should not be part of this PR

File: src/java_bytecode/expr2java.cpp [hunk 3]
```diff
@@ -235,6 +237,23 @@ std::string expr2javat::convert_constant(
     dest+='L';
     return dest;
   }
+  else if (src.type() == java_float_type()) {
```
Comment (mgudemann) [inline src/java_bytecode/expr2java.cpp]: could you please also run the `scripts/cpplint.py` script on this
For example, there should be no spaces around `=` or `==` or after `if`

File: src/java_bytecode/expr2java.cpp [hunk 4]
```diff
@@ -235,6 +237,23 @@ std::string expr2javat::convert_constant(
...
+    ieee_floatt Fpn(to_constant_expr(src));
```
Comment (peterschrammel) [inline src/java_bytecode/expr2java.cpp]: Do not use upper case in identifiers.

File: src/java_bytecode/expr2java.cpp [hunk 5]
```diff
@@ -235,6 +237,23 @@ std::string expr2javat::convert_constant(
...
+    float F = Fpn.to_float();
+    std::stringstream buffer;
+    buffer << std::hexfloat;
+    buffer << F << 'f';
+    return(buffer.str());
```
Comment (peterschrammel) [inline src/java_bytecode/expr2java.cpp]: `return buffer.str();` (several occurrences)

File: src/java_bytecode/expr2java.cpp [hunk 6]
```diff
@@ -235,6 +237,23 @@ std::string expr2javat::convert_constant(
     dest+='L';
     return dest;
   }
+  else if (src.type() == java_float_type()) {
+    ieee_floatt Fpn(to_constant_expr(src));
+    float F = Fpn.to_float();
+    std::stringstream buffer;
+    buffer << std::hexfloat;
+    buffer << F << 'f';
```
Comment (peterschrammel) [inline src/java_bytecode/expr2java.cpp]: You could merge this with the previous statement to make it shorter

**Comment by eigold** [comment]
Since this is my first CBMC patch,  Matthias asked me to submit a PR with
the code I added just to start the process.  (To tell the truth I thought
that 'in progress' meant the same thing as "don't disturb" on a hotel door.
Nevertheless your comments are greatly appreciated. ) This patch still
misses  the code that takes care about special cases e.g. NaN, infinity and
so on reported in 'ieee_floatt'.  After I add it, I will start working on
your comments.

**Comment by mgudemann** [comment]
@eigold no worries, for the "in progress", this means basically that you are working on it and yes, "ready to Review" will be the next step. But I have to admit that this is not always treated as it should be :-)

File: src/goto-programs/interpreter.cpp (hunk 1 above)
Comment (peterschrammel) [inline src/goto-programs/interpreter.cpp]: Please remove code instead of commenting out

**Comment by eigold** [comment]
Sorry, I have been distracted by work on a new ebmc release.
As for the code in question I will actually restore the commented-out line.
The problem is that  I run 32-bit Ubuntu.  Line
     'return 2ULL << 32ULL'

causes the error message below:

interpreter.cpp:964:20: error: large integer implicitly truncated to
unsigned type [-Werror=overflow]

Matthias suggested that I should  use a replacement line   'return 1ULL <<
31ULL'  on my computer. Since this line does not look good for 64-bit
systems, currently it is just my "personal patch" used to build a binary on
my desktop.

**Comment by peterschrammel** [comment]
Then this should be fixed in the interpreter to work on 32bit systems (can be done as a separate PR).

**Comment by eigold** [comment]
Okay

**Comment by mgudemann** [comment]
Please note that such a PR exists already but is currently not worked on
as 32 bit compatibility is not high priority for test-gen.

https://github.com/diffblue/cbmc/pull/987

The above fix is problematic as it will not allow creation of many
objects, and even for 32 bits 64 bit adressing should be used. But this
requires a more complex rewrite.

**Comment by mgudemann** [comment]
@eigold could you please change the PR, so that it does not contain the change to `interpreter.cpp`
This can be done by `editing` the commit and only staging/committing the other changes. This should be described under "interactive rebase" in more git resources. After then, you can `push --force` the changed commit.
You can also undo the change, commit that change and then squash the commits, again via interactive rebase.

**Comment by peterschrammel** [comment]
@eigold, any progress on this?

**Comment by eigold** [comment]
Yes. I have made the change  as a separate commit. It shows when
I  run 'git log' on my computer but for some reason github shows only
the first commit for this branch.
Eugene

**Comment by mgudemann** [comment]
@eigold as one of the last commits says "starting a new branch" could it be that the newer commit are simply in a different, new branch? Only commits pushed to `eigold:feature/hexadecimal_float_output_in_expr2java` will appear here

**Comment by eigold** [comment]
When I run 'git branch'  in ~/test-gen/lib/cbmc it shows
* feature/hexadecimal_float_output_in_expr2java
  master

The  last two commits that I get  after running 'git log'
in ~/test-gen/lib/cbmc are shown below.


Date:   Mon Jul 3 05:54:01 2017 -0400

    removing an accidental change

commit e46a320de0eaf8c99c4dd485448a25a764d6986d
Author: eigold <eu.goldberg@gmail.com>
Date:   Mon Jun 26 16:05:48 2017 -0400

    starting a new branch to use hexadecimal floating point numbers in
tests (amended)

commit 05875e1d7f5bc3d00b060d31bf79b9f57127d6a1

**Comment by mgudemann** [comment]
@eigold all these commits are shown on github then. You just have to scroll down, "removing and accidental change" is just right in front of Peter's last comment

**Comment by eigold** [comment]
Oops. You are right. I got distracted with the work on ebmc release and
forgot that  I gave the same names to the branches of test-gen and
a fork of cbmc. I was actually checking the status of the former that
indeed has not changed. Sorry.

Eugene

File: src/java_bytecode/expr2java.cpp [hunk 7]
```diff
@@ -237,23 +237,40 @@ std::string expr2javat::convert_constant(
...
+    {
+      std::string result;
+      if(ieee_repr.get_sign())
+        result="-";
+      else
+        result="+";
+      if(ieee_repr.is_NaN())
+        result+="NaN";
+      else
+        result+="inf";
```
Comment (mgudemann) [inline src/java_bytecode/expr2java.cpp]: for Java, this is `Double/Float.NEGATIVE_INFINITY` or `Double/Float.POSITIVE_INFINITY`
(cf. https://docs.oracle.com/javase/7/docs/api/java/lang/Float.html)

File: src/java_bytecode/expr2java.cpp [hunk 8]
```diff
@@ -237,23 +237,40 @@ std::string expr2javat::convert_constant(
...
+    ieee_floatt ieee_repr(to_constant_expr(src));
+    if(ieee_repr.is_NaN() || ieee_repr.is_infinity())
+    {
+      std::string result;
+      if(ieee_repr.get_sign())
+        result="-";
+      else
+        result="+";
+      if(ieee_repr.is_NaN())
+        result+="NaN";
```
Comment (mgudemann) [inline src/java_bytecode/expr2java.cpp]: for Java, this is `Float/Double.NaN`

File: src/java_bytecode/expr2java.cpp [hunk 9]
```diff
@@ -237,23 +237,40 @@ std::string expr2javat::convert_constant(
...
+    double d_num;
+    if(float_flag)
+      f_num=ieee_repr.to_float();
+    else
+      d_num=ieee_repr.to_double();
     std::stringstream buffer;
     buffer << std::hexfloat;
-    buffer << D;
+    if(float_flag)
+      buffer << f_num << 'f';
```
Comment (mgudemann) [inline src/java_bytecode/expr2java.cpp]: please simplify to `buffer << ieee_repr.to_float() << 'f';`

File: src/java_bytecode/expr2java.cpp [hunk 10]
```diff
@@ -237,23 +237,40 @@ std::string expr2javat::convert_constant(
...
+    else
+      buffer << d_num;
```
Comment (mgudemann) [inline src/java_bytecode/expr2java.cpp]: as above

File: src/java_bytecode/expr2java.cpp [hunk 11]
```diff
@@ -237,23 +237,40 @@ std::string expr2javat::convert_constant(
...
+    float f_num;
+    double d_num;
+    if(float_flag)
+      f_num=ieee_repr.to_float();
+    else
+      d_num=ieee_repr.to_double();
     std::stringstream buffer;
     buffer << std::hexfloat;
-    buffer << D;
+    if(float_flag)
```
Comment (mgudemann) [inline src/java_bytecode/expr2java.cpp]: nitpick, explicit `bool` is not needed, `if(src.type()==java_float_type())` would be enough

**Comment by eigold** [comment]
Matthias,

. . .
> please simplify to buffer << ieee_repr.to_float() << 'f';
...

I simplified the code as you suggested.

I missed the point of  references to Java functions
for checking NaN and Infinity.

**Comment by mgudemann** [comment]
Hi Eugene,

> I missed the point of  references to Java functions
> for checking NaN and Infinity.

this function emits Java code, therefore it should emit valid Java.

```
float f = Float.POSITIVE_INFINITY;
```
is valid, while
```
float f = +inf;
```
is not, but this would currently be generated.

These special numbers are all defines as static members of either
`java.lang.Float` or `java.lang.Double`.

Best,
Matthias

**Comment by eigold** [comment]
I see.

Note that this has nothing to do with outputting floating point numbers
in the hexadecimal form. I commented out the code I added and ran java
programs where  one has to go beyond regular floating point numbers to
reach some lines.
The test generator produced tests where value +NAN, -NAN, -INFINITY,
+INFINITY were assigned  (that is tests that would not compile).

So, this PR should count as two :-)

**Comment by mgudemann** [comment]
Hi Eugene,


> Note that this has nothing to do with outputting floating point numbers
> in the hexadecimal form. I commented out the code I added and ran java
> programs where  one has to go beyond regular floating point numbers to
> reach some lines.
> The test generator produced tests where value +NAN, -NAN, -INFINITY,
> +INFINITY were assigned  (that is tests that would not compile).

I fear I cannot follow you here.

What I did was to checkout your branch in `lib/cbmc` and run
test-generator on a test that requires NaN to reach a location. What I
got was the following test-case

```
/*
 * This test covers `boolean f(float)' block 2 (line 4)
 */

@org.junit.Test
public void nan_tst_f_a5f1433ea6d903e6_001() throws Throwable {

  boolean retval;
  {
    /* initialize test parameters */
    nan_tst param_1 = (nan_tst) Reflector.getInstance("nan_tst");
    float arg1f = -NaN;

    /* call function under test */
    retval = param_1.f(arg1f);
  }
  {
    /* check return value */
    java.lang.Object expected_return_value;
    expected_return_value = true;

    CompareWithFieldList.compare(expected_return_value, retval);
  }
}
```

Which is not valid Java.

I think ew should have `arg1f = -Float.NaN;` here. It is possible to
give hex values for +-inf/nan too, but that would not make much sense.

I'd also prefer to have `0` if zero is exactly represented (via
`is_zero` as you mentioned earlier).

Best,
Matthias

**Comment by eigold** [comment]
Right. And if you comment out the code I have added you should get

float arg1f = -NAN;

(instead of float arg1f = -NaN; )

That is currently (i.e. even without the patch I have added) test-gen
generates incorrect values  for java tests if a non-number or Infinity is
involved.
For instance, it assigns -NAN instead of Float.NaN.
That is all I am saying.

Eugene






On Thu, Jul 6, 2017 at 10:23 AM, Matthias Güdemann <notifications@github.com

**Comment by mgudemann** [comment]
> Right. And if you comment out the code I have added you should get
>
> float arg1f = -NAN;

which is because it falls back to expr2c, right? Which is using the NAN
macro as output for NaN.

> (instead of float arg1f = -NaN; )
>
> That is currently (i.e. even without the patch I have added) test-gen
> generates incorrect values  for java tests if a non-number or Infinity is
> involved.

yes, which is why we want your PR to correct this problem :-)

I'd say now mostly the only necessary change is from "NaN" to
"Float.NaN" or "Double.NaN" and do the case distinction for
"Double/Float.NEGATIVE/POSITIVE_INFINITY", then is should all be
set. Other hexadecimal float output should be ok with using `std::hex`.

Best,
Matthias

**Comment by eigold** [comment]
Probably, it makes sense to add three more regression tests to cover
all three special cases:   non-number, infinity,  and clean zero?


On Thu, Jul 6, 2017 at 10:53 AM, Matthias Güdemann <notifications@github.com

**Comment by eigold** [comment]
Done.

On Wed, Jul 12, 2017 at 3:37 PM, Matthias Güdemann <notifications@github.com

**Comment by mgudemann** [comment]
obsoleted by #1216
